### PR TITLE
Enhanced utils.py and webdriver.py

### PR DESCRIPTION
**Pull request by dmytrokoren**
1. Enhanced utils.py: Employed randomized timing for multiple attempts, mitigating 403 errors.
2. Revamped webdriver.py: Incorporated random timing during login to prevent 429 errors.

Tested on:
- Windows 11 docker, WSL2
- Synology DSM 7.2 (Container Manager)
- Mac OS docker, arm64
- Mac OS native, arm64

**NOTE: You have to be on your residential IP address (including docker), VPN will throw 403 on fare check!**

### CODE REVIEW

**Comment by Bubba8291** [comment]
@dmytrokoren This replaces #268 right? And what's different about it?

**Comment by dmytrokoren** [comment]
> @dmytrokoren This replaces #268 right? And what's different about it?

@Bubba8291 No difference, just minor tweaks.

**Review by jdholtz — COMMENTED** [review]
Thanks for adding and testing with the random intervals. I unfortunately have still been receiving some 403s using your `develop6` tag, but it has definitely improved over v7.5.

**Comment by dmytrokoren** [comment]
> Thanks for adding and testing with the random intervals. I unfortunately have still been receiving some 403s using your `develop6` tag, but it has definitely improved over v7.5.

I haven't pushed the latest changes yet. These changes don't fix 403 yet. I will push latest tomorrow. Please check latest tag not develop6 (it's outdated) thanks @jdholtz

File: lib/utils.py [hunk 1]
```diff
@@ -52,7 +56,9 @@ def make_request(method: str, site: str, headers: JSON, info: JSON, max_attempts
             logger.debug("Reservation not found")
             break
 
-        time.sleep(0.5)
+        attempts_sleep = set_sleep_duration()
+        logger.debug(f"Attempt {attempts}: Sleeping for {attempts_sleep:.2f} seconds")
```
[the remark this answers is not in the document]
Comment (dmytrokoren) [reply lib/utils.py:60]: Yes you are right we should separate fare check sleep from check-in. Totally agree

[the remark this answers is not in the document]
Comment (dmytrokoren) [reply lib/utils.py]: I modified this line of code, now it's only sleeping on fare check when going for multiple attempts.

File: lib/webdriver.py [hunk 2]
```diff
@@ -60,6 +62,9 @@ def __init__(self, checkin_scheduler: CheckInScheduler) -> None:
         self.login_status_code = None
         self.trips_request_id = None
 
+    def _set_sleep_duration(self) -> None:
+        return random.uniform(1, 5)
```
[the remark this answers is not in the document]
Comment (dmytrokoren) [reply lib/webdriver.py]: Modified this as well, now it's using it from utils with different parameters.

**Comment by dmytrokoren** [comment]
> Thanks for adding and testing with the random intervals. I unfortunately have still been receiving some 403s using your `develop6` tag, but it has definitely improved over v7.5.

@jdholtz - please try :latest tag

**Comment by jdholtz** [comment]
I've tried this on my machine, a Pi, and an Ubuntu server. They all work well. The only issue I'm running into is I'm occasionally getting the following error. It seems to happen more consistently on the Ubuntu server using Google Chrome instead of Chromium, but I did get it as well on the Pi for one run
```
seleniumbase.common.exceptions.ElementNotVisibleException: Message: 
 Element {(//button[contains(@class,'closeButton')])[2]} was not visible after 7 seconds!
```
I saw through screenshots that the popup did not appear, so I tried ignoring it. However, then the headers never were loaded. This PR is very close to fixing the issue, but I'm not exactly sure why this particular issue exists.

File: lib/reservation_monitor.py [hunk 3]
```diff
@@ -101,24 +101,38 @@ def _check_flight_fares(self) -> None:
             try:
                 fare_checker.check_flight_price(flight)
                 self.notification_handler.healthchecks_success(
-                    f"Successful fare check, confirmation number={flight.confirmation_number}"
```
Comment (jdholtz) [inline lib/reservation_monitor.py:104]: Is there a reason all of these were changed? It is just for healthchecks and the one-liner doesn't seem like it would be that hard to read--especially if this is only to make sure the fare checker is running.

**Comment by dmytrokoren** [comment]
> I've tried this on my machine, a Pi, and an Ubuntu server. They all work well. The only issue I'm running into is I'm occasionally getting the following error. It seems to happen more consistently on the Ubuntu server using Google Chrome instead of Chromium, but I did get it as well on the Pi for one run
> 
> ```
> 
> seleniumbase.common.exceptions.ElementNotVisibleException: Message: 
> 
>  Element {(//button[contains(@class,'closeButton')])[2]} was not visible after 7 seconds!
> 
> ```
> 
> I saw through screenshots that the popup did not appear, so I tried ignoring it. However, then the headers never were loaded. This PR is very close to fixing the issue, but I'm not exactly sure why this particular issue exists.

That is awesome news! I will fix this issue tonight and push changes.

File: lib/reservation_monitor.py [hunk 4]
```diff
@@ -101,24 +101,38 @@ def _check_flight_fares(self) -> None:
...
                 )
 
+    def _sleep_message(self, hours, minutes, seconds):
+        if self.first_name is not None and self.last_name is not None:
+            name = f" for {self.first_name} {self.last_name}"
+        else:
+            name = ""
+        
+        if hours > 0:
+            print(f"Sleeping for {hours} hours and {minutes} minutes until the next check{name}\n")
```
Comment (jdholtz) [inline lib/reservation_monitor.py:129]: I'm not a big fan of having this message printed. What's the reason for it? It's good for a debug message (which was already printed on line 130 before these changes). However, I don't think this isn't particularly useful to users. They already know it will check once per day and when this runs for enough time, the stdout will be filled with `Sleeping for 23 hours and 59 minutes until the next check for ...`.

File: lib/utils.py [hunk 5]
```diff
@@ -14,13 +15,16 @@
 JSON = Dict[str, Any]
 
 BASE_URL = "https://mobile.southwest.com/api/"
-NTP_SERVER = "us.pool.ntp.org"
```
Comment (jdholtz) [inline lib/utils.py:17]: Although this isn't directly relevant to the PR, this is a good change. Small and non-relevant changes would be best in a different PR so this one is easier to see exactly what changed to fix the 403s and 429s.

File: lib/webdriver.py [hunk 6]
```diff
@@ -130,7 +129,7 @@ def get_reservations(self, account_monitor: AccountMonitor) -> List[JSON]:
         return reservations
 
     def _get_driver(self) -> Driver:
-        logger.debug("Starting webdriver for current session (this may take a few minutes)")
```
Comment (jdholtz) [inline lib/webdriver.py:133]: Why was this removed? It was added due to #237 and put in both places (this and `Loading Southwest page`) as it may take a while at both places. The message `(this may take a moment)` has better wording than this though.

File: lib/webdriver.py [hunk 7]
```diff
@@ -201,6 +202,7 @@ def _wait_for_login(self, driver: Driver, account_monitor: AccountMonitor) -> No
             raise error
 
         self._set_account_name(account_monitor, login_response)
+        driver.click("div.nav-item.my-account-nav-item")
```
Comment (jdholtz) [inline lib/webdriver.py:205]: Does this help the script stay undetected and fetch the reservations correctly or is this unnecessary?

File: lib/webdriver.py [hunk 8]
```diff
@@ -247,8 +249,7 @@ def _handle_login_error(self, response: JSON) -> LoginError:
     def _get_needed_headers(self, request_headers: JSON) -> JSON:
         headers = {}
         for header in request_headers:
-            if re.match(r"x-api-key|x-channel-id|user-agent|^[\w-]+?-\w$", header, re.I):
-                headers[header] = request_headers[header]
+            headers[header] = request_headers[header]
```
Comment (jdholtz) [inline lib/webdriver.py:252]: Do more headers need to be matched? If all of the headers are saved, this function isn't needed. However, it's unclear to me whether or not all request headers are actually relevant to the script bypassing 403s.

File: lib/reservation_monitor.py (hunk 3 above)
Comment (dmytrokoren) [reply lib/reservation_monitor.py:104]: Slight appearance change from this:
IMG_63771F0694C4-1

to look like:
'
Successful fare check,
confirmation = 24K3XS
'

File: lib/reservation_monitor.py (hunk 4 above)
Comment (dmytrokoren) [reply lib/reservation_monitor.py:129]: We can change this back to logger.debug but I wanted to have some visual when not using debug to show when the script is finished checking and is on standby. As an example, if we have 3 reservations in total (running in accounts or in reservation mode) after the last reseravtion to indicate script is not entering standby because if we don't run in debug it does shows checked in for the following flights and that's it. That's just my opinion, what do you think?

File: lib/utils.py (hunk 5 above)
Comment (dmytrokoren) [reply lib/utils.py:17]: Ill make this change in another PR.

File: lib/webdriver.py (hunk 6 above)
Comment (dmytrokoren) [reply lib/webdriver.py:133]: As many tests I have done :) for me it took only a while for the Loading Southwest ... page
The  "Starting webdriver for current session (this may take a few minutes)" always loaded fast. But we can keep that message there as well, I just thought it's not needed.

**Comment by dmytrokoren** [comment]
@jdholtz check this tag :develop

File: lib/reservation_monitor.py (hunk 3 above)
Comment (jdholtz) [reply lib/reservation_monitor.py:104]: Gotcha, yeah that looks better. Thanks

File: lib/webdriver.py (hunk 6 above)
Comment (jdholtz) [reply lib/webdriver.py:133]: Yeah, that's been my experience too. It can be left off to just say `Starting webdriver for current session`.

File: lib/reservation_monitor.py (hunk 4 above)
Comment (jdholtz) [reply lib/reservation_monitor.py]: > after the last reservation to indicate script is now entering standby because if we don't run in debug it only shows checked in for the following flights and that's it.

Could you expand on the situation here? I don't quite follow.

The script will show the messages for when all reservations are scheduled to check in, but I'm not sure if users would find it very helpful to print out a message every day on every account/reservation to indicate that it has just checked it (if desired, that can be received through healthchecks as well as the log). If you and/or others do find this helpful, though, I'm fine with keeping it as a print statement (but still logging the `Sleeping for ... seconds` debug message that was removed). 

If it's just going to be a logger.debug message, I think it's fine to leave it as the `Sleeping for ... seconds` instead of having multiple branches for a debug message. It's easy enough to parse while debugging and not super helpful to have hours and minutes printed out versus just seconds (only within the debug message).

File: lib/reservation_monitor.py (hunk 3 above)
Comment (dmytrokoren) [reply lib/reservation_monitor.py:104]: Southwest uses 'confirmation #: 3KXSSA'

Should I make a change to be:
"
Successful fare check,
confirmation #: 3KXSSA
"

File: lib/reservation_monitor.py (hunk 4 above)
Comment (dmytrokoren) [reply lib/reservation_monitor.py]: I agree with you not to print for every account and every reservation. I was thinking to print only once at the end - "Sleeping for xxx seconds" after all accounts and reservations have been completed.

**Comment by jdholtz** [comment]
> @jdholtz check this tag :develop

I just pulled this tag. However, I'm getting the following error when running it through Docker on my laptop. I think the last strategy worked well (the MY_ACCOUNT_URL), but we just need to figure why it isn't always loading. However, it may be because it is being detected which would mean a new strategy is needed.
```
selenium.common.exceptions.ElementClickInterceptedException: Message: element click intercepted: Element <img alt="Check in banner" class="image-placement--background-image" height="51" src="/content/mkt/images/hero_shots/mobileSliver-CheckIn-test-20210330.png" width="390"> is not clickable at point (195, 280). Other element would receive the click: <h3 class="popup-title">...</h3>
  (Session info: chrome=125.0.6422.112)
```

There is another strategy that may be worth experimenting with which is running SeleniumBase with a virtual display: https://github.com/seleniumbase/SeleniumBase/issues/2390#issuecomment-1871385366. I believe I tried this a few months ago and it didn't work, but it's worth trying again in case I didn't do it correctly or a SB update improved it. 

In my opinion, the best way to fix 403s is to find out why they are happening on headless environments (docker, ubuntu server, etc.), for the most part, but not happening on devices with displays. I've spent a lot of time since the issue started to try and find out why but have not been successful yet.

File: lib/reservation_monitor.py (hunk 3 above)
Comment (jdholtz) [reply lib/reservation_monitor.py:104]: `confirmation number = ...` is good. It doesn't need to be 100% consistent with Southwest's website and I think this is easier to understand from a glance.

**Comment by dmytrokoren** [comment]
> > @jdholtz check this tag :develop
> 
> 
> 
> I just pulled this tag. However, I'm getting the following error when running it through Docker on my laptop. I think the last strategy worked well (the MY_ACCOUNT_URL), but we just need to figure why it isn't always loading. However, it may be because it is being detected which would mean a new strategy is needed.
> 
> ```
> 
> selenium.common.exceptions.ElementClickInterceptedException: Message: element click intercepted: Element Check in banner is not clickable at point (195, 280). Other element would receive the click: <h3 class="popup-title">...</h3>
> 
>   (Session info: chrome=125.0.6422.112)
> 
> ```
> 
> 
> 
> There is another strategy that may be worth experimenting with which is running SeleniumBase with a virtual display: https://github.com/seleniumbase/SeleniumBase/issues/2390#issuecomment-1871385366. I believe I tried this a few months ago and it didn't work, but it's worth trying again in case I didn't do it correctly or a SB update improved it. 
> 
> 
> 
> In my opinion, the best way to fix 403s is to find out why they are happening on headless environments (docker, ubuntu server, etc.), for the most part, but not happening on devices with displays. I've spent a lot of time since the issue started to try and find out why but have not been successful yet.

That's so weird. It worked on my docker Mac m1. I will need to use js_click that will avoid this error. I'll fix it again tonight :)

File: lib/reservation_monitor.py (hunk 4 above)
Comment (jdholtz) [reply lib/reservation_monitor.py]: The only problem with that is that every account/reservation can sleep for a different period (retrieval_interval can be configured on a per-account and per-reservation basis) so there's not only one correct message. The best solution may be to print the one with the lowest number of seconds but there is no communication between processes so I don't see that being possible (or even just one of these messages, as you suggested, instead of one per process)

Comment (dmytrokoren) [reply lib/reservation_monitor.py]: When user is not using debug... We don't need to display the seconds. Just at the end when everything is done to display "Monitoring is complete until the next check" something like that. When not running in debug right now it does tell you if it's still processing or it's done.

**Comment by dmytrokoren** [comment]
I will check to see if we can utilize the virtual desktop but I did start to poke at this but like you said unsuccessful :/

**Comment by dmytrokoren** [comment]
> > @jdholtz check this tag :develop
> 
> I just pulled this tag. However, I'm getting the following error when running it through Docker on my laptop. I think the last strategy worked well (the MY_ACCOUNT_URL), but we just need to figure why it isn't always loading. However, it may be because it is being detected which would mean a new strategy is needed.
> 
> ```
> selenium.common.exceptions.ElementClickInterceptedException: Message: element click intercepted: Element Check in banner is not clickable at point (195, 280). Other element would receive the click: <h3 class="popup-title">...</h3>
>   (Session info: chrome=125.0.6422.112)
> ```
> 
> There is another strategy that may be worth experimenting with which is running SeleniumBase with a virtual display: [seleniumbase/SeleniumBase#2390 (comment)](https://github.com/seleniumbase/SeleniumBase/issues/2390#issuecomment-1871385366). I believe I tried this a few months ago and it didn't work, but it's worth trying again in case I didn't do it correctly or a SB update improved it.
> 
> In my opinion, the best way to fix 403s is to find out why they are happening on headless environments (docker, ubuntu server, etc.), for the most part, but not happening on devices with displays. I've spent a lot of time since the issue started to try and find out why but have not been successful yet.

I have tried the virtual display (sbvirtualdisplay) it's not working for me, I received 429 error on login. The real issue here is user-agent it's detecting somehow. If we set is_mobile=False in virtual display it will fail fare check every time and login session with 429 error intermittently. Right now is_mobile=True it's working, with proper workarounds. No need virtual display.

**Comment by Bubba8291** [comment]
@dmytrokoren How often are the errors now with your branch? I told my cousin I would get her a good boarding position on Friday. But I assured her that it's not guaranteed.

I am also suitable to run it on a stable residential internet, so no VPS.

**Comment by dmytrokoren** [comment]
> @dmytrokoren How often are the errors now with your branch? I told my cousin I would get her a good boarding position on Friday. But I assured her that it's not guaranteed.
> 
> 
> 
> I am also suitable to run it on a stable residential internet, so no VPS.

@Bubba8291 try this tag (develop) https://hub.docker.com/r/dmytrokoren/auto-southwest-check-in/ 

I'm running it without any issues on fare check or scheduled check in. It's been stable for me at least. Give it a go and report back here.

**Comment by Bubba8291** [comment]
> > @dmytrokoren How often are the errors now with your branch? I told my cousin I would get her a good boarding position on Friday. But I assured her that it's not guaranteed.
> > I am also suitable to run it on a stable residential internet, so no VPS.
> 
> @Bubba8291 try this tag (develop) https://hub.docker.com/r/dmytrokoren/auto-southwest-check-in/
> 
> I'm running it without any issues on fare check or scheduled check in. It's been stable for me at least. Give it a go and report back here.

@dmytrokoren I do not use Docker. If I fetch and pull dmytrokoren/auto-southwest-check-in:master, it's the same commits as your latest Docker develop tag, right?

**Comment by dmytrokoren** [comment]
> @dmytrokoren I do not use Docker. If I fetch and pull dmytrokoren/auto-southwest-check-in:master, it's the same commits as your latest Docker develop tag, right?

If using desktop, I'm not sure if it works on Windows platform. On Mac it works but make sure it does not go to sleep.

First, download the script onto your computer

git clone https://github.com/dmytrokoren/auto-southwest-check-in.git

**Comment by Bubba8291** [comment]
> > @dmytrokoren I do not use Docker. If I fetch and pull dmytrokoren/auto-southwest-check-in:master, it's the same commits as your latest Docker develop tag, right?
> 
> If using desktop, I'm not sure if it works on Windows platform. On Mac it works but make sure it does not go to sleep.
> 
> First, download the script onto your computer
> 
> git clone https://github.com/dmytrokoren/auto-southwest-check-in.git

Yep. Thank you. I already have it setup on a headless machine. I just wanted to make sure your docker tag was the same as your latest GitHub branch.

**Comment by dmytrokoren** [comment]
> Yep. Thank you. I already have it setup on a headless machine. I just wanted to make sure your docker tag was the same as your latest GitHub branch.

Yup it is. Keep us posted how it checks in )

**Comment by jdholtz** [comment]
> I have tried the virtual display (sbvirtualdisplay) it's not working for me, I received 429 error on login. The real issue here is user-agent it's detecting somehow. If we set is_mobile=False in virtual display it will fail fare check every time and login session with 429 error intermittently. Right now is_mobile=True it's working, with proper workarounds. No need virtual display.

Thanks for checking. Very helpful and good to know the virtual display won't make a difference.

Just tested through Docker on my local machine and my Pi. Also tried on my Ubuntu server VM. All worked perfectly and didn't even fail an attempt when running with an account and reservation at the same time. I'll ask a few more people to verify and we can get this merged.

This is a great PR, it brings the script back to full functionality. Thanks for all of your hard work on this!

**Comment by jdholtz** [comment]
The only issue I was getting was a 403 when it was fetching the reservation information within the checkin scheduler. What fixed it for me (I'm not 100% confident because I don't think I got another failed attempt) is to add `is_fare_checker=True` to the `make_request` call in the `_get_reservation_info` function.

I'll probably change it to default to making random delays between requests and turn it off within the checkin handler, but you don't need to worry about doing that for this PR.

**Comment by Bubba8291** [comment]
@dmytrokoren I would like to double check since it's not in the README yet. Which browser is better for your fork using headless Ubuntu Focal (20.04)? Google chrome or Chromium?

**Comment by Bubba8291** [comment]
@dmytrokoren Good news kind of. My cousin got A46. But it took 10 attempts to be successful. I think she might've gotten lucky since it took 14 extra seconds.

```2024-06-21 14:09:55 DEBUG Process-1:1[checkin_handler:155]: Attempting to check in
2024-06-21 14:09:55 DEBUG Process-1:1[checkin_handler:188]: Making GET request to check in
2024-06-21 14:09:56 DEBUG Process-1:1[utils:64]: Sleeping for 0.50 seconds after 1 attempts
2024-06-21 14:09:58 DEBUG Process-1:1[utils:64]: Sleeping for 0.50 seconds after 2 attempts
2024-06-21 14:09:59 DEBUG Process-1:1[utils:64]: Sleeping for 0.50 seconds after 3 attempts
2024-06-21 14:10:00 DEBUG Process-1:1[utils:64]: Sleeping for 0.50 seconds after 4 attempts
2024-06-21 14:10:01 DEBUG Process-1:1[utils:64]: Sleeping for 0.50 seconds after 5 attempts
2024-06-21 14:10:02 DEBUG Process-1:1[utils:64]: Sleeping for 0.50 seconds after 6 attempts
2024-06-21 14:10:02 DEBUG Process-1:1[utils:64]: Sleeping for 0.50 seconds after 7 attempts
2024-06-21 14:10:07 DEBUG Process-1:1[utils:64]: Sleeping for 0.50 seconds after 8 attempts
2024-06-21 14:10:12 DEBUG Process-1:1[utils:64]: Sleeping for 0.50 seconds after 9 attempts
2024-06-21 14:10:14 DEBUG Process-1:1[utils:47]: Successfully made request after 10 attempts
2024-06-21 14:10:14 DEBUG Process-1:1[checkin_handler:194]: Making POST request to check in
2024-06-21 14:10:16 DEBUG Process-1:1[utils:47]: Successfully made request after 1 attempts
2024-06-21 14:10:16 DEBUG Process-1:1[checkin_handler:165]: Successfully checked in after 1 attempts
2024-06-21 14:10:16 DEBUG Process-1:1[notification_handler:105]: Sending successful check-in notification...```

**Comment by dmytrokoren** [comment]
@Bubba8291 A46 is good, considering people pay for A15-A30 early bird. Regarding GET request was attempting multiple times, I think SW api was not responsive at that time as there is no 403 error. At this point there is very little that can be done related to this. My environment just like yours DSM, is working good right now with everything one attempt. I'm yet to try checkin, I have a trip coming up in mid July. I can report back how my check in went.

**Comment by dmytrokoren** [comment]
> @dmytrokoren I would like to double check since it's not in the README yet. Which browser is better for your fork using headless Ubuntu Focal (20.04)? Google chrome or Chromium?

I'm using Chromium headless. How it is setup in docker file, nothing changed.

**Comment by dmytrokoren** [comment]
> The only issue I was getting was a 403 when it was fetching the reservation information within the checkin scheduler. What fixed it for me (I'm not 100% confident because I don't think I got another failed attempt) is to add `is_fare_checker=True` to the `make_request` call in the `_get_reservation_info` function.
> 
> I'll probably change it to default to making random delays between requests and turn it off within the checkin handler, but you don't need to worry about doing that for this PR.

Added the random_sleep (prev aka is_fare_checker) to _get_reservation_info in checkin_scheduler.

> @dmytrokoren Good news kind of. My cousin got A46. But it took 10 attempts to be successful. I think she might've gotten lucky since it took 14 extra seconds.
> 
> ```
> 2024-06-21 14:09:55 DEBUG Process-1:1[checkin_handler:188]: Making GET request to check in
> 2024-06-21 14:09:56 DEBUG Process-1:1[utils:64]: Sleeping for 0.50 seconds after 1 attempts
> 2024-06-21 14:09:58 DEBUG Process-1:1[utils:64]: Sleeping for 0.50 seconds after 2 attempts
> 2024-06-21 14:09:59 DEBUG Process-1:1[utils:64]: Sleeping for 0.50 seconds after 3 attempts
> 2024-06-21 14:10:00 DEBUG Process-1:1[utils:64]: Sleeping for 0.50 seconds after 4 attempts
> 2024-06-21 14:10:01 DEBUG Process-1:1[utils:64]: Sleeping for 0.50 seconds after 5 attempts
> 2024-06-21 14:10:02 DEBUG Process-1:1[utils:64]: Sleeping for 0.50 seconds after 6 attempts
> 2024-06-21 14:10:02 DEBUG Process-1:1[utils:64]: Sleeping for 0.50 seconds after 7 attempts
> 2024-06-21 14:10:07 DEBUG Process-1:1[utils:64]: Sleeping for 0.50 seconds after 8 attempts
> 2024-06-21 14:10:12 DEBUG Process-1:1[utils:64]: Sleeping for 0.50 seconds after 9 attempts
> 2024-06-21 14:10:14 DEBUG Process-1:1[utils:47]: Successfully made request after 10 attempts
> 2024-06-21 14:10:14 DEBUG Process-1:1[checkin_handler:194]: Making POST request to check in
> 2024-06-21 14:10:16 DEBUG Process-1:1[utils:47]: Successfully made request after 1 attempts
> 2024-06-21 14:10:16 DEBUG Process-1:1[checkin_handler:165]: Successfully checked in after 1 attempts
> 2024-06-21 14:10:16 DEBUG Process-1:1[notification_handler:105]: Sending successful check-in notification...```
> ```

@jdholtz also added random_sleep to make_request in [checkin_handler: line 188] the reason is that @Bubba8291 had some api request issue and it only went through after 10 attempts.

btw random_sleep I have made a change instead of (1, 6) it is now (1, 3) so it have a high delay time.

File: lib/utils.py [hunk 9]
```diff
@@ -20,7 +21,10 @@
 RESERVATION_NOT_FOUND_CODE = 400620389
 
 
-def make_request(method: str, site: str, headers: JSON, info: JSON, max_attempts=20) -> JSON:
+def set_sleep_duration(min_duration: float, max_duration: float) -> float:
```
Comment (jdholtz) [inline lib/utils.py:24]: Could the name of this function be changed to `random_sleep_duration`? That way it makes more sense as it isn’t setting anything within it.

File: lib/checkin_handler.py [hunk 10]
```diff
@@ -186,7 +186,7 @@ def _check_in_to_flight(self) -> JSON:
         site = CHECKIN_URL + self.flight.confirmation_number
 
         logger.debug("Making GET request to check in")
-        response = make_request("GET", site, headers, info)
+        response = make_request("GET", site, headers, info, random_sleep=True)
```
Comment (jdholtz) [inline lib/checkin_handler.py:189]: I’d like someone to test with a check-in to see how much this will affect the time it takes (preferably on a device that almost never has 403s) to make sure this doesn’t make check-ins take too long to go through. Unfortunately I won’t have a flight to test for a while.

**Review by jdholtz — CHANGES_REQUESTED** [review]
Just a couple of small things before I merge this. Don’t worry about adding tests (the workflow will fail when I accept it). I’ll add them after the merge.

**Comment by jdholtz** [comment]
@dmytrokoren the linting/formatting workflow is failing. Could you fix those issues before I merge? 

As a tip, I recommend using pre-commit (details are in the developers doc) so the formatters automatically run and fix it for you.

**Comment by dmytrokoren** [comment]
> @dmytrokoren the linting/formatting workflow is failing. Could you fix those issues before I merge?
> 
> As a tip, I recommend using pre-commit (details are in the developers doc) so the formatters automatically run and fix it for you.

Done with pre-commit

black....................................................................Passed
codespell................................................................Passed
flake8...................................................................Passed
isort....................................................................Passed

**Comment by bradzab0623** [comment]
@jdholtz @dmytrokoren I've started using this version and it does seem pretty good. One thing I did notice is when trying to log in or check fares for many accounts and/or reservations at once, I get errors again after the first couple. Can we stagger the log ins and fare checks when running the script against many accounts and/or reservations? I understand we can't do that with the actual check ins.

**Comment by jdholtz** [comment]
> Can we stagger the log ins and fare checks when running the script against many accounts and/or reservations?

Yeah, it may work better if there was a stagger between each account/reservation. Because the script sleeps for exactly 24 hours (or whatever retrieval interval is set), this only needs to be implemented when each process is started as the times between each will stay consistent throughout the script's running time.

@dmytrokoren I have also been starting to see 403s with just one account and 2 reservations again. Very confused what could've changed...

File: lib/checkin_handler.py (hunk 10 above)
Comment (jdholtz) [reply lib/checkin_handler.py:189]: Although the second check-in didn't pass, [this comment](https://github.com/jdholtz/auto-southwest-check-in/issues/230#issuecomment-2192313132) shows that it doesn't seemingly affect the check-in time. More samples should be observed, but this verifies that it logically shouldn't make much of a difference. The check-in will still go through in the first 10ish seconds no matter if the script sleeps for 0.5 seconds or 1-3 seconds. From check-ins in the past, the fastest I've seen it is 5 seconds after the minute.

**Comment by dmytrokoren** [comment]
> > Can we stagger the log ins and fare checks when running the script against many accounts and/or reservations?
> 
> Yeah, it may work better if there was a stagger between each account/reservation. Because the script sleeps for exactly 24 hours (or whatever retrieval interval is set), this only needs to be implemented when each process is started as the times between each will stay consistent throughout the script's running time.
> 
> @dmytrokoren I have also been starting to see 403s with just one account and 2 reservations again. Very confused what could've changed...

@jdholtz I've been running 4 accounts / 0 reservations for 4 days now in docker (synology). Not one single 403/429 error (fingers crossed) - fare checks and header refresh working flawlessly. That's why im confused why its happening to other people?!

File: lib/main.py [hunk 11]
```diff
@@ -87,8 +87,8 @@ def set_up_check_in(arguments: List[str]) -> None:
         logger.error("Invalid arguments. For more information, try '--help'")
         sys.exit(2)
 
-    logger.debug(
-        "Monitoring %d accounts and %d reservations", len(config.accounts), len(config.reservations)
+    print(
```
Comment (evanchen7) [inline lib/main.py:90]: Did you mean to use print instead of logger.debug?

**Comment by dmytrokoren** [comment]
@jdholtz is the PR going to be merged?

File: lib/main.py (hunk 11 above)
Comment (jdholtz) [reply lib/main.py:90]: I think it was meant to be print. It's helpful for debugging though so I'll make it a `logger.info`.

**Comment by jdholtz** [comment]
Yes, just did @dmytrokoren. Thanks!

**Comment by natecodes** [comment]
```
2024-07-01 12:12:44 DEBUG MainProcess[log:24]: Initialized the application
2024-07-01 12:12:44 DEBUG MainProcess[main:113]: Auto-Southwest Check-In v7.5
2024-07-01 12:12:44 DEBUG MainProcess[main:71]: Called with 0 arguments
2024-07-01 12:12:44 DEBUG MainProcess[config:132]: Initializing configuration file
2024-07-01 12:12:44 DEBUG MainProcess[config:161]: Reading the configuration file
2024-07-01 12:12:44 DEBUG MainProcess[config:174]: Reading configuration from environment variables
2024-07-01 12:12:44 DEBUG MainProcess[config:60]: Setting check fares to True
2024-07-01 12:12:44 DEBUG MainProcess[config:75]: Setting retrieval interval to 24 hours
2024-07-01 12:12:44 DEBUG MainProcess[config:109]: Setting notification level to <NotificationLevel.INFO: 1>
2024-07-01 12:12:44 DEBUG MainProcess[config:122]: Using 0 notification services
2024-07-01 12:12:44 DEBUG MainProcess[config:144]: Creating configurations for 1 accounts
2024-07-01 12:12:44 DEBUG MainProcess[config:151]: Creating configurations for 0 reservations
2024-07-01 12:12:44 INFO MainProcess[main:97]: Monitoring 1 account and 0 reservations

2024-07-01 12:12:44 DEBUG Process-1[reservation_monitor:168]: Acquiring lock...
2024-07-01 12:12:44 DEBUG Process-1[reservation_monitor:170]: Lock acquired
2024-07-01 12:12:44 DEBUG Process-1[reservation_monitor:192]: Retrieving reservations for account
2024-07-01 12:12:44 DEBUG Process-1[webdriver:126]: Starting webdriver for current session
2024-07-01 12:12:45 DEBUG Process-1[webdriver:143]: Using browser version: 126.0.6478.127
2024-07-01 12:12:45 DEBUG Process-1[webdriver:147]: Loading Southwest home page (this may take a moment)
2024-07-01 12:12:52 DEBUG Process-1[webdriver:100]: Logging into account to get a list of reservations and valid headers
2024-07-01 12:12:54 DEBUG Process-1[webdriver:177]: Waiting for headers_set to be set
2024-07-01 12:12:54 DEBUG Process-1[webdriver:181]: headers_set set successfully
2024-07-01 12:12:55 DEBUG Process-1[webdriver:177]: Waiting for login_request_id to be set
2024-07-01 12:12:55 DEBUG Process-1[webdriver:169]: Login response has been received
2024-07-01 12:12:56 DEBUG Process-1[webdriver:181]: login_request_id set successfully
2024-07-01 12:12:56 DEBUG Process-1[webdriver:237]: Logging in failed for an unknown reason
2024-07-01 12:12:56 WARNING Process-1[reservation_monitor:201]: Encountered a Too Many Requests error while logging in. Skipping reservation retrieval
2024-07-01 12:12:56 DEBUG Process-1[reservation_monitor:181]: Lock released
2024-07-01 12:12:56 DEBUG Process-1[reservation_monitor:130]: Sleeping for 86387 seconds
^C2024-07-01 12:12:58 INFO MainProcess[main:122]:
Ctrl+C pressed. Stopping all check-ins
```
